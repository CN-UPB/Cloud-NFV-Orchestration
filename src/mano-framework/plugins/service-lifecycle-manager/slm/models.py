from datetime import datetime
from uuid import uuid4

from mongoengine import Document, EmbeddedDocument
from mongoengine.fields import (
    DateTimeField,
    DictField,
    EmbeddedDocumentListField,
    StringField,
    UUIDField,
)


class Function(EmbeddedDocument):
    """
    Represents a function instance.
    """

    id = UUIDField(required=True)
    descriptor = DictField(required=True)
    vim = UUIDField()
    record = DictField()


class Service(Document):
    """
    Represents a service instance.
    """

    id = UUIDField(primary_key=True, default=uuid4)
    created_at = DateTimeField(default=datetime.utcnow)

    status = StringField()

    descriptor = DictField(required=True)
    functions = EmbeddedDocumentListField(Function, required=True)

    placement = DictField()
