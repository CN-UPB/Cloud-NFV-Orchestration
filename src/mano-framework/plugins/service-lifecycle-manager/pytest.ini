[pytest]
# Make pytest-env set the ENV variable to "test" so the proper configuration
# file is loaded
env =
    ENV=test