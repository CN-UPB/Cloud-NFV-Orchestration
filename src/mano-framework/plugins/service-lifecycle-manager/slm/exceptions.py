class DeployRequestValidationError(Exception):
    def __init__(self, message: str):
        super().__init__(message)


class InstantiationError(Exception):
    def __init__(self, message: str):
        super().__init__(message)

    def add_error(self, e: Exception):
        """
        Stringify a provided exception and add it to the error message of the
        instantiation error.
        """
        self.message += "\n" + str(e)


class TerminationError(Exception):
    def __init__(self, message: str):
        super().__init__(message)


class PlacementError(InstantiationError):
    def __init__(self):
        super().__init__("Unable to perform placement.")
