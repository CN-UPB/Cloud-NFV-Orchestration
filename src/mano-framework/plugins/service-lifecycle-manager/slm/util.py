from time import time
from typing import Type, Union


def create_status_message(
    status: str = None, error: Union[str, Exception] = None, payload={}
):
    if status is None:
        if error is not None:
            status = "ERROR"
        else:
            status = "SUCCESS"

    if isinstance(error, Exception):
        error = str(error)

    return {
        **payload,
        "status": status,
        "error": error,
        "timestamp": time(),
    }


def get_vm_image_id(vnfd: dict, vdu: dict):
    return "{}_{}_{}_{}".format(
        vnfd["vendor"], vnfd["name"], vnfd["version"], vdu["id"]
    )


def raise_on_ia_error(
    ia_response: dict, exception_class: Type[Exception], logger=None, *log_args
):
    """
    Given a response message from the infrastructure adaptor, raises an exception of the
    specified exception class if the response indicates an error. The error message of
    the IA is provided as the only argument to the constructor of the exception class.
    If `logger` is provided, its `info` method will be invoked with `log_args` if
    `ia_response` indicates an error.

    Args:
        ia_response: The response from the IA
        exception_class: The exception class to be used in case of an error
        logger: An optional logger instance to log the error
        *log_args: The arguments that will be provided to the `info` method of `logger`
    """
    if ia_response["request_status"] == "ERROR":
        message = ia_response["message"]
        if logger is not None:
            logger.info(*log_args)
            logger.debug("Error: %s", message)
        raise exception_class(message)
