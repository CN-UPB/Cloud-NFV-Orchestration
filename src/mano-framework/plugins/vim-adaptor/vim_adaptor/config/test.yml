mongo: mongomock://localhost/vim-adaptor
