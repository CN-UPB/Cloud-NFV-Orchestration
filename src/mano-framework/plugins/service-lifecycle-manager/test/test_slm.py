from pathlib import Path
from test.util import simple_async_endpoint
from uuid import uuid4

import pytest
import yaml

from manobase.messaging import AsyncioBrokerConnection as Connection
from manobase.messaging import Message
from slm import topics
from slm.exceptions import InstantiationError, PlacementError, TerminationError
from slm.slm import ServiceLifecycleManager

TEST_DIR = Path(__file__).parent
DESCRIPTORS_DIR = TEST_DIR / "descriptors"


def load_descriptor(filename: str):
    with (DESCRIPTORS_DIR / filename).with_suffix(".yml").open() as f:
        return yaml.safe_load(f)


SERVICE_DESCRIPTOR = load_descriptor("sonata-demo")
FUNCTION_DESCRIPTORS = [
    load_descriptor(f) for f in ["firewall-vnfd", "iperf-vnfd", "tcpdump-vnfd"]
]

# A static VIM ID so it can be used in snapshot testing
VIM_ID = "87ecdc7c-665a-475a-b4f7-f053706ceb70"

PLACEMENT = {function["uuid"]: {"vim": VIM_ID} for function in FUNCTION_DESCRIPTORS}


@pytest.fixture
def deploy_request():
    """
    The payload of a deploy request message like the gatekeeper sends it.
    """
    return {"nsd": SERVICE_DESCRIPTOR, "vnfds": FUNCTION_DESCRIPTORS}


@pytest.fixture
def topology(connection: Connection):
    """
    Mocks the Infrastructure Adapter's response to topology requests and yields the
    topology that the IA returns
    """
    topology = []

    with simple_async_endpoint(connection, topics.IA_TOPOLOGY, topology):
        yield topology


@pytest.fixture
def mocked_placement(connection: Connection):
    """
    Mocks the placement plugin's response to placement requests with the global
    `PLACEMENT` constant
    """
    with simple_async_endpoint(connection, topics.MANO_PLACE, {"mapping": PLACEMENT}):
        yield


@pytest.fixture
async def fetched_placement(manager: ServiceLifecycleManager, mocked_placement):
    """
    Makes the manager retrieve a placement using `manager._fetch_placement()` and the
    `mocked_placement` fixture
    """
    await manager._fetch_placement(topology=[])


@pytest.fixture(scope="function")
def manager(connection: Connection, mongo_connection, deploy_request):
    """
    A fully initialized ServiceLifecycleManager instance
    """

    manager = ServiceLifecycleManager.from_deploy_request(
        Message(
            topic=topics.MANO_DEPLOY,
            payload=deploy_request,
            correlation_id=str(uuid4()),
        ),
        connection,
    )

    # Fix the service instance id for snapshot tests
    manager.service.id = "f86d5ec7-1a93-404d-80d1-e696369f8318"
    manager.service.save()

    yield manager

    manager.service.delete()  # Delete the Service document from MongoDB


@pytest.mark.asyncio
async def test_fetch_topology(manager: ServiceLifecycleManager, topology):
    assert topology == await manager._fetch_topology()


@pytest.mark.asyncio
async def test_fetch_placement(
    manager: ServiceLifecycleManager, connection: Connection, snapshot_endpoint
):
    topic = topics.MANO_PLACE

    # Should send placement request that matches snapshot
    with snapshot_endpoint(topic, {"mapping": PLACEMENT}):
        await manager._fetch_placement(topology=[])

    # Should store the placement in the Service document and its Function documents
    assert PLACEMENT == manager.service.placement
    for function in manager.service.functions:
        assert str(function.vim) == VIM_ID

    # Should raise PlacementError if the returned mapping is ``None``
    with simple_async_endpoint(connection, topic, {"mapping": None}):
        with pytest.raises(PlacementError):
            await manager._fetch_placement(topology=[])


@pytest.mark.asyncio
async def test_prepare_infrastructure(
    manager: ServiceLifecycleManager,
    connection: Connection,
    fetched_placement,
    snapshot_endpoint,
):
    topic = topics.IA_PREPARE

    # Should send request that matches snapshot
    with snapshot_endpoint(topic, response={"request_status": "COMPLETED"}):
        await manager._prepare_infrastructure()

    # Should raise errors from the IA
    with simple_async_endpoint(
        connection, topic, {"request_status": "ERROR", "message": "failed"}
    ):
        with pytest.raises(InstantiationError, match="failed"):
            await manager._prepare_infrastructure()


@pytest.mark.asyncio
async def test_deploy_vnfs(
    manager: ServiceLifecycleManager,
    connection: Connection,
    fetched_placement,
    snapshot_endpoint,
):
    topic = topics.MANO_DEPLOY

    # Should send requests that match their snapshots
    with snapshot_endpoint(
        topic, response={"request_status": "COMPLETED", "vnfr": {"key": "value"}}
    ):
        await manager._deploy_vnfs()

    # Should raise errors from the IA
    with simple_async_endpoint(
        connection, topic, {"request_status": "ERROR", "message": "failed"}
    ):
        with pytest.raises(InstantiationError, match="failed"):
            await manager._deploy_vnfs()


@pytest.mark.asyncio
async def test_destroy_vnfs(
    manager: ServiceLifecycleManager, connection: Connection, snapshot_endpoint,
):
    topic = topics.IA_REMOVE

    # Should send request that matches snapshot
    with snapshot_endpoint(topic, response={"request_status": "COMPLETED"}):
        await manager._destroy_vnfs()

    # Should raise errors from the IA
    with simple_async_endpoint(
        connection, topic, {"request_status": "ERROR", "message": "failed"}
    ):
        with pytest.raises(TerminationError, match="failed"):
            await manager._destroy_vnfs()
