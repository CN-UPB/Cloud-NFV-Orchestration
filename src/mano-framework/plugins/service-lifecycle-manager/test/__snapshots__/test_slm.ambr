# name: test_deploy_vnfs
  <class 'dict'> {
    'descriptor': <class 'BaseDict'> {
      'author': 'Steven van Rossem, iMinds',
      'connection_points': <class 'BaseList'> [
        <class 'dict'> {
          'id': 'mgmt',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'input',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'output',
          'type': 'interface',
        },
      ],
      'description': '
        "A first firewall VNF descriptor"
  
      ',
      'descriptor_version': 'vnfd-schema-01',
      'monitoring_rules': <class 'BaseList'> [
        <class 'dict'> {
          'condition': 'vdu01:vm_cpu_perc > 10',
          'description': 'Trigger events if CPU load is above 10 percent.',
          'duration': 10,
          'duration_unit': 's',
          'name': 'mon:rule:vm_cpu_perc',
          'notification': <class 'list'> [
            <class 'dict'> {
              'name': 'notification01',
              'type': 'rabbitmq_message',
            },
          ],
        },
        <class 'dict'> {
          'condition': 'vdu01:vm_mem_perc > 10',
          'description': 'Trigger events if memory consumption is above 10 percent.',
          'duration': 10,
          'duration_unit': 's',
          'name': 'mon:rule:vm_mem_perc',
          'notification': <class 'list'> [
            <class 'dict'> {
              'name': 'notification02',
              'type': 'rabbitmq_message',
            },
          ],
        },
        <class 'dict'> {
          'condition': 'vdu01:vm_net_rx_MB > 10',
          'duration': 10,
          'duration_unit': 's',
          'name': 'mon:rule:rx',
          'notification': <class 'list'> [
            <class 'dict'> {
              'name': 'notification03',
              'type': 'rabbitmq_message',
            },
          ],
        },
        <class 'dict'> {
          'condition': 'vdu01:vm_net_tx_MB > 10',
          'duration': 10,
          'duration_unit': 's',
          'name': 'mon:rule:tx',
          'notification': <class 'list'> [
            <class 'dict'> {
              'name': 'notification04',
              'type': 'rabbitmq_message',
            },
          ],
        },
      ],
      'name': 'firewall-vnf',
      'uuid': '6a15313f-cb0a-4540-baa2-77cc6b3f5b68',
      'vendor': 'eu.sonata-nfv',
      'version': '0.2',
      'virtual_deployment_units': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points': <class 'list'> [
            <class 'dict'> {
              'id': 'vdu01:eth0',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:eth1',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:eth2',
              'type': 'interface',
            },
          ],
          'id': 'vdu01',
          'monitoring_parameters': <class 'list'> [
            <class 'dict'> {
              'name': 'vm_cpu_perc',
              'unit': 'Percentage',
            },
            <class 'dict'> {
              'name': 'vm_mem_perc',
              'unit': 'Percentage',
            },
            <class 'dict'> {
              'name': 'vm_net_rx_MB',
              'unit': 'MB',
            },
            <class 'dict'> {
              'name': 'vm_net_tx_MB',
              'unit': 'Mbps',
            },
          ],
          'resource_requirements': <class 'dict'> {
            'cpu': <class 'dict'> {
              'vcpus': 1,
            },
            'memory': <class 'dict'> {
              'size': 2,
              'size_unit': 'GB',
            },
            'storage': <class 'dict'> {
              'size': 10,
              'size_unit': 'GB',
            },
          },
          'vm_image': 'http://registry.sonata-nfv.eu/html/files/VM_images/sonata-VM-2ports.qcow',
          'vm_image_format': 'qcow2',
        },
      ],
      'virtual_links': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:eth0',
            'mgmt',
          ],
          'connectivity_type': 'E-LAN',
          'dhcp': True,
          'id': 'mgmt',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:eth1',
            'input',
          ],
          'connectivity_type': 'E-Line',
          'dhcp': True,
          'id': 'input',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:eth2',
            'output',
          ],
          'connectivity_type': 'E-Line',
          'dhcp': True,
          'id': 'output',
        },
      ],
    },
    'id': '6a15313f-cb0a-4540-baa2-77cc6b3f5b68',
    'service_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
    'vim': '87ecdc7c-665a-475a-b4f7-f053706ceb70',
  }
---
# name: test_deploy_vnfs.1
  <class 'dict'> {
    'descriptor': <class 'BaseDict'> {
      'author': 'Steven van Rossem, iMinds',
      'connection_points': <class 'BaseList'> [
        <class 'dict'> {
          'id': 'mgmt',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'input',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'output',
          'type': 'interface',
        },
      ],
      'description': '
        "A first iperf VNF descriptor. The iperf VNF
         acts as a traffic source."
  
      ',
      'descriptor_version': 'vnfd-schema-01',
      'name': 'iperf-vnf',
      'uuid': '645db4fa-a714-4cba-9617-4001477d1281',
      'vendor': 'eu.sonata-nfv',
      'version': '0.2',
      'virtual_deployment_units': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points': <class 'list'> [
            <class 'dict'> {
              'id': 'vdu01:cp01',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:cp02',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:cp03',
              'type': 'interface',
            },
          ],
          'id': 'vdu01',
          'resource_requirements': <class 'dict'> {
            'cpu': <class 'dict'> {
              'vcpus': 1,
            },
            'memory': <class 'dict'> {
              'size': 2,
              'size_unit': 'GB',
            },
            'storage': <class 'dict'> {
              'size': 10,
              'size_unit': 'GB',
            },
          },
          'vm_image': 'file:///docker_files/iperf/Dockerfile',
          'vm_image_format': 'docker',
        },
      ],
      'virtual_links': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp01',
            'mgmt',
          ],
          'connectivity_type': 'E-LAN',
          'id': 'mgmt',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp02',
            'input',
          ],
          'connectivity_type': 'E-Line',
          'id': 'input',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp03',
            'output',
          ],
          'connectivity_type': 'E-Line',
          'id': 'output',
        },
      ],
    },
    'id': '645db4fa-a714-4cba-9617-4001477d1281',
    'service_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
    'vim': '87ecdc7c-665a-475a-b4f7-f053706ceb70',
  }
---
# name: test_deploy_vnfs.2
  <class 'dict'> {
    'descriptor': <class 'BaseDict'> {
      'author': 'Steven van Rossem, iMinds',
      'connection_points': <class 'BaseList'> [
        <class 'dict'> {
          'id': 'mgmt',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'input',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'output',
          'type': 'interface',
        },
      ],
      'description': '
        "A first tcpdump VNF descriptor. The tcpdump
         VNF acts as a traffic sink."
  
      ',
      'descriptor_version': 'vnfd-schema-01',
      'name': 'tcpdump-vnf',
      'uuid': '8a0aa837-ec1c-44e5-9907-898f6401c3ae',
      'vendor': 'eu.sonata-nfv',
      'version': '0.2',
      'virtual_deployment_units': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points': <class 'list'> [
            <class 'dict'> {
              'id': 'vdu01:cp01',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:cp02',
              'type': 'interface',
            },
            <class 'dict'> {
              'id': 'vdu01:cp03',
              'type': 'interface',
            },
          ],
          'id': 'vdu01',
          'resource_requirements': <class 'dict'> {
            'cpu': <class 'dict'> {
              'vcpus': 1,
            },
            'memory': <class 'dict'> {
              'size': 2,
              'size_unit': 'GB',
            },
            'storage': <class 'dict'> {
              'size': 10,
              'size_unit': 'GB',
            },
          },
          'vm_image': 'file:///docker_files/tcpdump/Dockerfile',
          'vm_image_format': 'docker',
        },
      ],
      'virtual_links': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp01',
            'mgmt',
          ],
          'connectivity_type': 'E-LAN',
          'id': 'mgmt',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp02',
            'input',
          ],
          'connectivity_type': 'E-Line',
          'id': 'input',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vdu01:cp03',
            'output',
          ],
          'connectivity_type': 'E-Line',
          'id': 'output',
        },
      ],
    },
    'id': '8a0aa837-ec1c-44e5-9907-898f6401c3ae',
    'service_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
    'vim': '87ecdc7c-665a-475a-b4f7-f053706ceb70',
  }
---
# name: test_destroy_vnfs
  <class 'dict'> {
    'service_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
  }
---
# name: test_fetch_placement
  <class 'dict'> {
    'functions': <class 'list'> [
      <class 'BaseDict'> {
        'author': 'Steven van Rossem, iMinds',
        'connection_points': <class 'BaseList'> [
          <class 'dict'> {
            'id': 'mgmt',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'input',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'output',
            'type': 'interface',
          },
        ],
        'description': '
          "A first firewall VNF descriptor"
  
        ',
        'descriptor_version': 'vnfd-schema-01',
        'monitoring_rules': <class 'BaseList'> [
          <class 'dict'> {
            'condition': 'vdu01:vm_cpu_perc > 10',
            'description': 'Trigger events if CPU load is above 10 percent.',
            'duration': 10,
            'duration_unit': 's',
            'name': 'mon:rule:vm_cpu_perc',
            'notification': <class 'list'> [
              <class 'dict'> {
                'name': 'notification01',
                'type': 'rabbitmq_message',
              },
            ],
          },
          <class 'dict'> {
            'condition': 'vdu01:vm_mem_perc > 10',
            'description': 'Trigger events if memory consumption is above 10 percent.',
            'duration': 10,
            'duration_unit': 's',
            'name': 'mon:rule:vm_mem_perc',
            'notification': <class 'list'> [
              <class 'dict'> {
                'name': 'notification02',
                'type': 'rabbitmq_message',
              },
            ],
          },
          <class 'dict'> {
            'condition': 'vdu01:vm_net_rx_MB > 10',
            'duration': 10,
            'duration_unit': 's',
            'name': 'mon:rule:rx',
            'notification': <class 'list'> [
              <class 'dict'> {
                'name': 'notification03',
                'type': 'rabbitmq_message',
              },
            ],
          },
          <class 'dict'> {
            'condition': 'vdu01:vm_net_tx_MB > 10',
            'duration': 10,
            'duration_unit': 's',
            'name': 'mon:rule:tx',
            'notification': <class 'list'> [
              <class 'dict'> {
                'name': 'notification04',
                'type': 'rabbitmq_message',
              },
            ],
          },
        ],
        'name': 'firewall-vnf',
        'uuid': '6a15313f-cb0a-4540-baa2-77cc6b3f5b68',
        'vendor': 'eu.sonata-nfv',
        'version': '0.2',
        'virtual_deployment_units': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points': <class 'list'> [
              <class 'dict'> {
                'id': 'vdu01:eth0',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:eth1',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:eth2',
                'type': 'interface',
              },
            ],
            'id': 'vdu01',
            'monitoring_parameters': <class 'list'> [
              <class 'dict'> {
                'name': 'vm_cpu_perc',
                'unit': 'Percentage',
              },
              <class 'dict'> {
                'name': 'vm_mem_perc',
                'unit': 'Percentage',
              },
              <class 'dict'> {
                'name': 'vm_net_rx_MB',
                'unit': 'MB',
              },
              <class 'dict'> {
                'name': 'vm_net_tx_MB',
                'unit': 'Mbps',
              },
            ],
            'resource_requirements': <class 'dict'> {
              'cpu': <class 'dict'> {
                'vcpus': 1,
              },
              'memory': <class 'dict'> {
                'size': 2,
                'size_unit': 'GB',
              },
              'storage': <class 'dict'> {
                'size': 10,
                'size_unit': 'GB',
              },
            },
            'vm_image': 'http://registry.sonata-nfv.eu/html/files/VM_images/sonata-VM-2ports.qcow',
            'vm_image_format': 'qcow2',
          },
        ],
        'virtual_links': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:eth0',
              'mgmt',
            ],
            'connectivity_type': 'E-LAN',
            'dhcp': True,
            'id': 'mgmt',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:eth1',
              'input',
            ],
            'connectivity_type': 'E-Line',
            'dhcp': True,
            'id': 'input',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:eth2',
              'output',
            ],
            'connectivity_type': 'E-Line',
            'dhcp': True,
            'id': 'output',
          },
        ],
      },
      <class 'BaseDict'> {
        'author': 'Steven van Rossem, iMinds',
        'connection_points': <class 'BaseList'> [
          <class 'dict'> {
            'id': 'mgmt',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'input',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'output',
            'type': 'interface',
          },
        ],
        'description': '
          "A first iperf VNF descriptor. The iperf VNF
           acts as a traffic source."
  
        ',
        'descriptor_version': 'vnfd-schema-01',
        'name': 'iperf-vnf',
        'uuid': '645db4fa-a714-4cba-9617-4001477d1281',
        'vendor': 'eu.sonata-nfv',
        'version': '0.2',
        'virtual_deployment_units': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points': <class 'list'> [
              <class 'dict'> {
                'id': 'vdu01:cp01',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:cp02',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:cp03',
                'type': 'interface',
              },
            ],
            'id': 'vdu01',
            'resource_requirements': <class 'dict'> {
              'cpu': <class 'dict'> {
                'vcpus': 1,
              },
              'memory': <class 'dict'> {
                'size': 2,
                'size_unit': 'GB',
              },
              'storage': <class 'dict'> {
                'size': 10,
                'size_unit': 'GB',
              },
            },
            'vm_image': 'file:///docker_files/iperf/Dockerfile',
            'vm_image_format': 'docker',
          },
        ],
        'virtual_links': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp01',
              'mgmt',
            ],
            'connectivity_type': 'E-LAN',
            'id': 'mgmt',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp02',
              'input',
            ],
            'connectivity_type': 'E-Line',
            'id': 'input',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp03',
              'output',
            ],
            'connectivity_type': 'E-Line',
            'id': 'output',
          },
        ],
      },
      <class 'BaseDict'> {
        'author': 'Steven van Rossem, iMinds',
        'connection_points': <class 'BaseList'> [
          <class 'dict'> {
            'id': 'mgmt',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'input',
            'type': 'interface',
          },
          <class 'dict'> {
            'id': 'output',
            'type': 'interface',
          },
        ],
        'description': '
          "A first tcpdump VNF descriptor. The tcpdump
           VNF acts as a traffic sink."
  
        ',
        'descriptor_version': 'vnfd-schema-01',
        'name': 'tcpdump-vnf',
        'uuid': '8a0aa837-ec1c-44e5-9907-898f6401c3ae',
        'vendor': 'eu.sonata-nfv',
        'version': '0.2',
        'virtual_deployment_units': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points': <class 'list'> [
              <class 'dict'> {
                'id': 'vdu01:cp01',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:cp02',
                'type': 'interface',
              },
              <class 'dict'> {
                'id': 'vdu01:cp03',
                'type': 'interface',
              },
            ],
            'id': 'vdu01',
            'resource_requirements': <class 'dict'> {
              'cpu': <class 'dict'> {
                'vcpus': 1,
              },
              'memory': <class 'dict'> {
                'size': 2,
                'size_unit': 'GB',
              },
              'storage': <class 'dict'> {
                'size': 10,
                'size_unit': 'GB',
              },
            },
            'vm_image': 'file:///docker_files/tcpdump/Dockerfile',
            'vm_image_format': 'docker',
          },
        ],
        'virtual_links': <class 'BaseList'> [
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp01',
              'mgmt',
            ],
            'connectivity_type': 'E-LAN',
            'id': 'mgmt',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp02',
              'input',
            ],
            'connectivity_type': 'E-Line',
            'id': 'input',
          },
          <class 'dict'> {
            'connection_points_reference': <class 'list'> [
              'vdu01:cp03',
              'output',
            ],
            'connectivity_type': 'E-Line',
            'id': 'output',
          },
        ],
      },
    ],
    'nap': <class 'dict'> {
    },
    'nsd': <class 'BaseDict'> {
      'author': 'Michael Bredel, NEC Labs Europe',
      'connection_points': <class 'BaseList'> [
        <class 'dict'> {
          'id': 'ns:mgmt',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'ns:input',
          'type': 'interface',
        },
        <class 'dict'> {
          'id': 'ns:output',
          'type': 'interface',
        },
      ],
      'description': '
        "The network service descriptor for the SONATA demo,
         comprising iperf, a firewall, and tcpdump."
  
      ',
      'descriptor_version': '2.0',
      'forwarding_graphs': <class 'BaseList'> [
        <class 'dict'> {
          'constituent_virtual_links': <class 'list'> [
            'mgmt',
            'input-2-iperf',
            'iperf-2-firewall',
            'firewall-2-tcpdump',
            'tcpdump-2-output',
          ],
          'constituent_vnfs': <class 'list'> [
            'vnf_iperf',
            'vnf_firewall',
            'vnf_tcpdump',
          ],
          'fg_id': 'ns:fg01',
          'network_forwarding_paths': <class 'list'> [
            <class 'dict'> {
              'connection_points': <class 'list'> [
                <class 'dict'> {
                  'connection_point_ref': 'ns:input',
                  'position': 1,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_iperf:input',
                  'position': 2,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_iperf:output',
                  'position': 3,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_firewall:input',
                  'position': 4,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_firewall:output',
                  'position': 5,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_tcpdump:input',
                  'position': 6,
                },
                <class 'dict'> {
                  'connection_point_ref': 'vnf_tcpdump:output',
                  'position': 7,
                },
                <class 'dict'> {
                  'connection_point_ref': 'ns:output',
                  'position': 8,
                },
              ],
              'fp_id': 'ns:fg01:fp01',
              'policy': 'none',
            },
          ],
          'number_of_endpoints': 2,
          'number_of_virtual_links': 4,
        },
      ],
      'name': 'sonata-demo',
      'network_functions': <class 'BaseList'> [
        <class 'dict'> {
          'vnf_id': 'vnf_firewall',
          'vnf_name': 'firewall-vnf',
          'vnf_vendor': 'eu.sonata-nfv',
          'vnf_version': '0.1',
        },
        <class 'dict'> {
          'vnf_id': 'vnf_iperf',
          'vnf_name': 'iperf-vnf',
          'vnf_vendor': 'eu.sonata-nfv',
          'vnf_version': '0.1',
        },
        <class 'dict'> {
          'vnf_id': 'vnf_tcpdump',
          'vnf_name': 'tcpdump-vnf',
          'vnf_vendor': 'eu.sonata-nfv',
          'vnf_version': '0.1',
        },
      ],
      'uuid': '005606ed-be7d-4ce3-983c-847039e3a5a2',
      'vendor': 'eu.sonata-nfv.service-descriptor',
      'version': '0.2',
      'virtual_links': <class 'BaseList'> [
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vnf_iperf:mgmt',
            'vnf_firewall:mgmt',
            'vnf_tcpdump:mgmt',
            'ns:mgmt',
          ],
          'connectivity_type': 'E-LAN',
          'id': 'mgmt',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'ns:input',
            'vnf_iperf:input',
          ],
          'connectivity_type': 'E-Line',
          'id': 'input-2-iperf',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vnf_iperf:output',
            'vnf_firewall:input',
          ],
          'connectivity_type': 'E-Line',
          'id': 'iperf-2-firewall',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vnf_firewall:output',
            'vnf_tcpdump:input',
          ],
          'connectivity_type': 'E-Line',
          'id': 'firewall-2-tcpdump',
        },
        <class 'dict'> {
          'connection_points_reference': <class 'list'> [
            'vnf_firewall:output',
            'ns:output',
          ],
          'connectivity_type': 'E-Line',
          'id': 'tcpdump-2-output',
        },
      ],
    },
    'serv_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
    'topology': <class 'list'> [
    ],
  }
---
# name: test_prepare_infrastructure
  <class 'dict'> {
    'instance_id': 'f86d5ec7-1a93-404d-80d1-e696369f8318',
    'vims': <class 'list'> [
      <class 'dict'> {
        'id': '87ecdc7c-665a-475a-b4f7-f053706ceb70',
        'vm_images': <class 'list'> [
          <class 'list'> [
            <class 'dict'> {
              'id': 'eu.sonata-nfv_firewall-vnf_0.2_vdu01',
              'md5': None,
              'url': 'http://registry.sonata-nfv.eu/html/files/VM_images/sonata-VM-2ports.qcow',
            },
          ],
          <class 'list'> [
            <class 'dict'> {
              'id': 'eu.sonata-nfv_iperf-vnf_0.2_vdu01',
              'md5': None,
              'url': 'file:///docker_files/iperf/Dockerfile',
            },
          ],
          <class 'list'> [
            <class 'dict'> {
              'id': 'eu.sonata-nfv_tcpdump-vnf_0.2_vdu01',
              'md5': None,
              'url': 'file:///docker_files/tcpdump/Dockerfile',
            },
          ],
        ],
      },
    ],
  }
---
